python/cudf/cudf/core/column/datetime.py

```diff
@@ -251,12 +252,16 @@ def binary_operator(self, op, rhs, reflect=False):
     def fillna(self, fill_value):
         if is_scalar(fill_value):
             if not isinstance(fill_value, Scalar):
-                fill_value = np.datetime64(fill_value, self.time_unit)
+                fill_value = cudf.Scalar(fill_value, dtype=self.dtype)
         else:
             fill_value = column.as_column(fill_value, nan_as_null=False)
 
         result = libcudf.replace.replace_nulls(self, fill_value)
-        if isinstance(fill_value, np.datetime64) and np.isnat(fill_value):
+        if (
+            isinstance(fill_value, cudf.Scalar)
+            and fill_value.dtype.type == np.datetime64
+            and np.isnat(fill_value.value)
```

Comment:
Any chance we could replace this with just `cudf.isnat(fill_value)`? This is not at all an intuitive way to check if a scalar is `NaT` and I'm wary of this pattern proliferating

Reply:
Do we even support a `NaT` value for datetime / timedelta scalars in cuDF? In Numpy / Pandas they use `min(int64)` as a sentinel value for this. Arrow doesn't have a concept of `NaT`, only `null`. I tend to believe we should follow Arrow.

Reply:
We have to treat NaT separately to deal with `fillna`. If we use just a null, it will drop the mask which is what we want to avoid. Filling with NaT keeps the mask but sets the underyling values to min(int64). 

[this is the behavior](https://github.com/rapidsai/cudf/blob/branch-0.17/python/cudf/cudf/core/column/datetime.py#L259-L264) that we want to maintain that causes the problem. We can move towards arrow in this respect but the tradeoff would be possibly having a harder time reproducing pandas behavior.

Reply:
Yea, if we pass in a numpy or pandas `NaT` scalar we should continue to maintain this behavior. The question though is if we support a `NaT` cudf scalar, which I believe we do not. Maybe we should set a flag somewhere earlier in this function or rework something downstream of this function instead?

Reply:
I see - so, no libcudf NaT -> no cuDF python NaT, that makes sense. Should be able to code around that - more to come here.

Reply:
@kkraus14 @shwina I looked into this will the help of @galipremsagar . I believe that while we don't have a `NaT` scalar in libcudf, it might be useful to support it in the python layer. 

Right now we actually *do* have a `NaT` `cudf._lib.scalar.Scalar`, i.e. the cdef class, that holds a true numpy nat object inside it, and eventually passes that numpy nat object directly into libcudf. It ends up working because numpy `NaT` is just a thin wrapper around `min(int64)`: 

```
>>> np.datetime64('NaT', 'ns').view('int64')
-9223372036854775808
```

Which is the number libcudf actually sees. So even though we don't have the concept of a true nat scalar in libcudf, the numpy nat object has a 1:1 correspondence with it for practical terms. To remove it would mean instead of passing that nat object directly to libcudf, we'd special case and pass an `int64` in those cases. Is that more natural than transparently supporting `nat`?

Reply:
After offline discussion with everyone, the plan for this is:

1. We will not support nat, and we will write the code such that NaT is treated as null for ops like `fillna` (e.g. a no-op)
2. Anything that relies on the current NaT behavior will rely on a separate codepath to special case around those scenarios